Add a custom source code formatter to replace `rust-fmt`

cuddlefishie opened the pull request:
Generated code looks very messy when not formatted,
`rustfmt` take a lot of time to format the code, so this crate
implements a reduced formatter that's supposed to run fast enough
that it can be part of the regular build process.


TODOs:
- [ ] change `StopWatch` so the timing summary makes sense again
- [ ] more "hard numbers" for the performance of the new formatter (MB/s?)

bluenote10 commented:
Just wondering, would [prettyplease](https://github.com/dtolnay/prettyplease) fit in here?

GodotRust commented:
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-303

Bromeon commented:
> Just wondering, would [prettyplease](https://github.com/dtolnay/prettyplease) fit in here?

Yes, we considered it, but it takes a `syn` syntax tree, not `TokenStream` -- so we'd need to add a heavy dependency and do full-blown Rust parsing. We also looked at [rust_format](https://github.com/nu11ptr/flexgen/tree/master/rust_format), but from what I can see, it's mostly an abstraction around rustfmt or prettyplease.

Bromeon commented on godot-codegen/src/lib.rs:
```diff
@@ -33,44 +34,68 @@ use proc_macro2::{Ident, TokenStream};
     watch: &mut godot_bindings::StopWatch,
 ) {
-    let mut out_files = vec![];
+    #[cfg(not(feature = "codegen-fmt"))]
+    let mut submit_fn = |path: PathBuf, tokens: TokenStream| {
+        write_file(&path, format_tokens(tokens));
+    };
+
+    #[cfg(feature = "codegen-fmt")]
+    let mut submit_fn = |path, tokens| {
+        write_file(&path, tokens.to_string());
+    };
```
Also, the two conditionals are swapped 🙂 

If feature `codegen-fmt` is enabled, then you should run `format_tokens`, otherwise not.
(I got an internal compiler error when swapping and incrementally compile further; had to clean).

If formatting turns out to have negligible performance impact, we can always enable it; but I'd suggest to do this outside of this PR.

cuddlefishie replied:
Oops, good catch, I put it there for testing 🙈 Corrected that! For enabling it by default, somebody should probably do proper measurements! For me it seems like it added about 200-400 ms in total.

Bromeon commented on godot-fmt/src/lib.rs:
```diff
@@ -0,0 +1,350 @@
+}
+
+fn indent(n: usize) -> &'static str {
+    // This looks strange but it means we don't need to actually allocate anything.
+    // The downside is there's a limit to how deep we can nest.
+    // The code that's generated doesn't seem like it's any deeper than this.
+
+    //                 |   |   |   |   |   |   |   |   |   |   |   |   |   |   |
+    let idents = "                                                        ";
+
+    let end = n * 4;
+    &idents[0..end]
```
Nice trick! 🙂

cuddlefishie replied:
Thanks 😅 Now with better formatting that doesn't depend on the type-inlay in vscode to properly align 😬

Bromeon commented on godot-fmt/src/lib.rs:
```diff
@@ -0,0 +1,350 @@
+/// "Best effort" here means:
+/// - line breaks are added in certain points
+///   - attributes
+///   - braces `{}`
+///   - semicolons `;`
+///   - colons if they follow braces `..},`
+/// - whitespace around punctuation is tried to be made as natural as possible
+///   - `::` always stick to whatever comes after them `use ::crate::module::{..}`
+///   - `$` is always prefix
+///   - `!` tries to join with whatever comes after it, `!boolean`, `x != y`
+///   - `&` tries to join with whatever comes after it, `&self`, `&'static`, `a && b`
+///   - joint punctuation tries to stay joint
+///   - if an identifier is followed by `.` or `()` then it's "sticky", `f()`, `s.x`
+///
+/// Current "flaws" (still compiles properly, just a bit ugly):
+/// - Because of stickiness, bitwise AND doesn't really look nice, like `a &mask`.
+///   Unless we check for all possible keywords (which would be costly), we
+///   can't solve this.
+/// - Also because of stickiness, macros are a bit weird, `name !`. If we made
+///   all `!` stick to previous identifiers then we'd get `x!= y`.
+///   Maybe that's a good trade-off if your code contains more not-equals checks
+///   than macro invocations.
+/// - anonymous function literals are a bit weird because the `|` don't have
+///   joint spacing when they come out of a `TokenStream` and identifiers don't
+///   have any spacing information attached, so it looks like `| arg |`.
+///   Also to know whether it's a chained bitwise or a closure would need more
+///   information. `move | arg | x`
+/// - Because we don't keep track of what's a type and what's an expression,
+///   generic parameters look a bit spaced out. `T < A, B >`
+/// - `match` arms can be `<pat> => <expr>,` and there _should_ be a line break
+///   after the `,` but because this only looks at one TokenKind at a time
+///   and only keeps a single state, it's too much effort to keep track of which
+///   `,` in a [`TokenGroup`] should trigger a newline (it would involve a stack
+///   of infos whether that group is in a match or not, detecting pattern etc)
+/// - Because `,` can be used for function args (no-newline) and separators
+///   in `struct` and `enum` definitons, those definitions are more awkward
+///   and crammed into one line as a trade-off to make function calls and args
+///   smoother.
+///   Maybe it's worth to make ALL args on their own line, then function calls
+///   get big, but `struct`s and `enum`s would be more natural?
```
Thanks so much for documenting this so detailed! ❤️

Bromeon commented on godot-fmt/Cargo.toml:
```diff
@@ -0,0 +1,16 @@
+
+[[bench]]
+name = "gdext_tests"
+harness = false
```
Should maybe have trailing newline.

Bromeon commented on godot-fmt/benches/gdext_tests.rs:
```diff
@@ -0,0 +1,34 @@
+        .unwrap()
+        .join("test-cases");
+
+    for rd in std::fs::read_dir(test_cases_dir).unwrap() {
+        let dir_entry = rd.unwrap();
+        let path = dir_entry.path();
+
+        let name = path.file_stem().unwrap();
+
+        let contents = std::fs::read_to_string(dir_entry.path()).unwrap();
+        let stream = TokenStream::from_str(&contents).unwrap();
+
+        c.bench_function(name.to_str().unwrap(), move |b| {
+            b.iter(|| format_tokens(black_box(stream.clone())))
+        });
+    }
```
Some comments:
* `rd` could just be named `file`, then you can do another `let file = file.unwrap();`
* variable `path` can be reused also inside `read_to_string`
* maybe `name` could be declared immediately before use:
  `let name = path.file_stem().unwrap().to_str().unwrap();`

cuddlefishie replied:
I named it `dir_entry` because it's technically not a file 🙈 

Did the other two things, nice catch, thanks!!

Bromeon commented on godot-fmt/benches/gdext_tests.rs:
```diff
@@ -0,0 +1,34 @@
+use proc_macro2::TokenStream;
+
+pub fn criterion_benchmark(c: &mut Criterion) {
+    // this little dance is needed because `cargo bench` and `cargo criterion`
+    // set the working directory slightly differently 🫠
+    let current_path = PathBuf::from(file!());
+    let test_cases_dir = current_path
+        .parent()
+        .unwrap()
+        .parent()
+        .unwrap()
+        .join("test-cases");
```
Unfortunately it doesn't seem to work for me with `cargo bench` on Windows 10:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value:
   Os { code: 3, kind: NotFound, message: "Cannot find specified path." }',
   godot-fmt\benches\gdext_tests.rs:18:4
```
(OS message translated)

I used `env!("CARGO_MANIFEST_DIR")` elsewhere, which seemed to work quite reliably. Maybe look for that in existing code and use it here too?

Bromeon commented on godot-fmt/src/lib.rs:
```diff
@@ -0,0 +1,350 @@
+    LastHash,
+}
+
+fn format(
+    mut state: FormatState,
+    level: usize,
+    tts: impl Iterator<Item = TokenTree>,
+    s: &mut String,
+) {
+    for tt in tts {
```
Would it make sense to have a `format_one` method accepting only a single `tt`? Then this function `format` could just loop over `format_one` with each element.

That way, the method doesn't need to be generic, and we can save one level of indentation instead of a giant `for` loop.

Bromeon replied:
Also, you seem to call `format` mostly with `group.stream().into_iter()` and once with `tokens.into_iter()`. Could that signature be made more specific to the call sites?

Bromeon commented on godot-fmt/src/lib.rs:
```diff
@@ -0,0 +1,350 @@
+// *prepend* whitespace or not.
+//
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum FormatState {
+    Start,
+    NothingSpecial,
+    LastJoinedOperator,
+    LastDoubleColon,
+
+    LastIdentifier,
+    LastClosingBrace,
+    LastHash,
+}
```
Could you briefly document those? Some are self-explanatory, but others like `NothingSpecial` aren't 🙂

Bromeon commented on godot-fmt/benches/gdext_tests.rs:
```diff
@@ -0,0 +1,34 @@
+use std::{path::PathBuf, str::FromStr};
```
Since this is a benchmark and not a test, would it make sense to name the file `gdext_bench.rs`?

Bromeon commented on godot-fmt/src/lib.rs:
```diff
@@ -0,0 +1,350 @@
+                        state = FormatState::LastDoubleColon;
+                    }
+                    (_, FormatState::Start) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
+                    (_, FormatState::NothingSpecial) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
+                    (_, FormatState::LastJoinedOperator) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
+                    (_, FormatState::LastDoubleColon) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
+                    ('.', FormatState::LastIdentifier) => {
+                        state = FormatState::LastJoinedOperator;
+                    }
+                    (_, FormatState::LastIdentifier) => {
+                        if punct.spacing() == Spacing::Joint {
+                            state = FormatState::LastJoinedOperator;
+                        } else {
+                            state = FormatState::NothingSpecial;
+                        }
+                    }
```
All these conditions except for `'.'`  have the same body, could you collapse the match arms with `|`?

Bromeon reviewed (commented):
Thanks so much for this! The output code looks really much nicer 🙂 

The `test-cases` directory is mostly for benchmarking, right? Do the 5 files differ substantially, or could we do with just 3 or so? (Asking because they add 0.5 MB to the repo forever; even if compressed in Git).

Also, I don't see the formatting time output in `*stats.txt` files, would it be possible to include that again?

cuddlefishie commented:
I addressed the review comments! Thanks!

Because of the way `StopWatch` is set up and the way the `SubmitFn` is performing the formatting (or not) immediately, it's harder to get direct timing.

Comparing the total time in `codegen-stats.txt` with and without formatting gave me inconsistent results. Sometimes running with formatting was *faster* (?!), sometimes the other way around. So if somebody else can test this, that would be good :)

Bromeon reviewed (commented):
Thanks a lot for this! The benchmarking and `StopWatch` measurements can be addressed at a later time.

bors r+
